support cluster option

topliceanu opened the pull request.

topliceanu commented:
CC @hamchapman

hamchapman commented:
It LGTM but probably good idea to get @zimbatm to take a look as well

topliceanu commented:
@zimbatm can you please do a code review on this PR. I'm concerned about messing up the logic in the initializer: https://github.com/pusher/pusher-http-ruby/blob/94-support-cluster-option/lib/pusher/client.rb#L12-L39

mdpye commented:
Is this our opportunity to deprecate the config globals as discussed in #48 ?

We'll be hitting the docs to add a call to `cluster=` everywhere, so we could change them to use an instance at the same time.

We don't have to remove the globals, but by not adding new ones we can deprecate them.

topliceanu commented:
@mdpye Oh yeah! I was wondering about that global connection attached to the `Pusher` namespace. I completely agree, we should deprecate it. We need to keep the docs for the global as well, maybe under a deprecated label.

zimbatm commented on README.md:
````diff
@@ -74,7 +80,18 @@ pusher_client = Pusher::Client.new({
+If you created your app in a cluster different from the default one, you must pass the `cluster` option like so:
+
+``` ruby
+pusher_client = Pusher::Client.new({
````
Don't put the `{}` they are unecessary

topliceanu replied:
They are used everywhere in this README, so for consistency I will keep them.

zimbatm replied:
Okay. This is bad ruby style, the lib needs an overhaul.

zimbatm commented on README.md:
````diff
@@ -43,6 +43,12 @@ By default API requests are made over HTTP. HTTPS can be used by setting
+If you created your app in a cluster different from the default one, you must pass the cluster option:
+
+``` ruby
+Pusher.cluster = 'your-app-cluster'
````
We want to deprecate the singleton approach. If the user wants to use the cluster option he should create a new instance of the Pusher::Client instead.

zimbatm commented on lib/pusher/client.rb:
```diff
@@ -88,6 +98,10 @@ def encrypted?
     end
 
+    def cluster=(cluster)
+      @host = "api-#{cluster}.pusher.com"
```
The `@port` should also be reset to `nil` in that case. We then depend on the `@scheme` to select the right port.

topliceanu replied:
I don't understand why we should set `@port=nil` when setting `@cluster`. cluster option only interacts with the host, not with the port. Why should we force the user to configure `@scheme` when configuring `@host` ?

zimbatm commented on pusher.gemspec:
```diff
@@ -3,7 +3,7 @@ $:.push File.expand_path("../lib", __FILE__)
 Gem::Specification.new do |s|
   s.name        = "pusher"
-  s.version     = "0.16.0"
+  s.version     = "0.17.0"
```
Making a release should be done independently from the PR.

The best way to do a release is as follows:

* bump the gemspec version
* run `bundle` if the Gemfile.lock is part of the repo
* run `git changelog` from the `git-extras` package to generate the changelog, edit for humans
* `git commit -a -m "Release v$version"`
* `git tag v$version; git push; git push --tags`
* `rm *.gem; gem build *.gemspec && gem push *.gem`

topliceanu commented:
- I updated the README. Now, using `Pusher::Client.new()` if the first option and the global `Pusher` is clearly marked as being deprecated. Could use a proof read. CC @zimbatm @mdpye [new readme]( https://github.com/pusher/pusher-http-ruby/blob/94-support-cluster-option/README.md)

- I made setting `cluster` consistent to all other libraries, ie. if the user sets the host or the url of the connection, then cluster is ignored. I also added tests for all this stuff.

zimbatm commented on lib/pusher/client.rb:
```diff
@@ -88,6 +105,12 @@ def encrypted?
       @scheme == 'https'
     end
 
+    def cluster=(cluster)
```
If the user explicitly calls `pusher.cluster = 'eu'` then we shouldn't ignore that. We should reset all the other conflicting settings.

mdpye replied:
Calling both is an error. Do we
 - raise?
 - define a clear precedence of one over the other?
 - just go with the last one to be called?

That's pretty much in my descending order of preference. Open to argument, obviously...

zimbatm replied:
Exactly, it raises to many questions. These options only exist because of the singleton and should be deprecated with it. If the user wants to change the config he can always create a new client and pass the options in the constructor.

Until then I think that the last option is the best, both other options need to record the ordering of independent method calls.

topliceanu replied:
I agree, I'll make the changes. Thanks for helping out!
